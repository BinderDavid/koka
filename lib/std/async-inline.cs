/*---------------------------------------------------------------------------
  Copyright 2017 Microsoft Corporation.
 
  This is free software; you can redistribute it and/or modify it under the
  terms of the Apache License, Version 2.0. A copy of the License can be
  found in the file "license.txt" at the root of this distribution.
---------------------------------------------------------------------------*/
using System.Threading;

static class _Async
{
  public class ThreadTimer {
    private Primitive.EventloopEntry entry;
    private Timer timer;

    public ThreadTimer( Fun0<Unit> cb, int ms ) {
      entry = Primitive.GetEventloopEntry();
      if (ms <= 0) {
        timer = null;
        entry.Post(() => { cb.Apply(); });
      }
      else {
        timer = new Timer( (object state0) => { 
                  if (entry != null) entry.Post(() => { cb.Apply(); }); 
                }, null, ms, Timeout.Infinite );
      }
    }

    public void Close() {
      if (entry != null) entry.Close();
      if (timer != null) timer.Dispose();
      entry = null;
      timer = null;
    }
  }

  public static ThreadTimer SetTimeout( Fun0<Unit> cb, int ms ) {
    return new ThreadTimer(cb,ms);
  }

  public static void ClearTimeout( object obj ) {
    if (obj == null) return;
    ThreadTimer t = (ThreadTimer)obj;
    t.Close();
  }
}
