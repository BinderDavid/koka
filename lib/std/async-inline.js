/*---------------------------------------------------------------------------
  Copyright 2012-2016 Microsoft Corporation.
 
  This is free software; you can redistribute it and/or modify it under the
  terms of the Apache License, Version 2.0. A copy of the License can be
  found in the file "license.txt" at the root of this distribution.
---------------------------------------------------------------------------*/

var effCancel = "std/async/cancelable";

function _create_cancelable( caction ) {
  // identity operations
  function ret( loc, x, _k ) { return _k(x); };
  function ops(_l1, _l2, _cont, _op, _resume , _k ) { return _resume(_l1, _l2, _cont, _unit_, _k);  };
  var handle = $std_core._makeHandler1( effCancel, ret, ops );
  var loc = { canceled:false, cancelers:[] };
  function action(_k) { return caction( function(){ return _cancel(loc); }, _k); }
  return function(_k){ return handle( loc, action, _k ); } 
}

function throw_canceled() {
  $std_core.throw_1("Canceled",Cancel);
}

function _cancel( loc ) {
  loc.canceled = true;
  var cs = loc.cancelers; loc.cancelers = [];
  cs.forEach( function(cb) { 
    if (typeof cb.cancel === "function") { 
      cb.cancel();
      setTimeout( function() {
        $std_core._restore_handler_context(cb.hctx, function(_) {
          throw_canceled(); // throw a Cancel exception in that context
        });
      }, 0);
    }
  });
  //return $std_core._unit_;
}

function _check_canceled( hctx ) {
  if (hctx==null) hctx = $std_core._save_handler_context();
  var top = hctx.htop;
  while (top >= 0) {
    var h = hctx.hstack[top];
    if (h.optag===effCancel && h.loc1.canceled) throw_canceled();
    top--;
  }
  //return $std_core._unit_;
}

function _register_cancel( cb, f ) {
  var hctx = $std_core._save_handler_context();
  var top = hctx.htop;
  while (top >= 0 && hctx.hstack[top].optag !== effCancel) { top--; }
  if (top >= 0) {
    //console.log("set cancel: " + cb + ", f: " + f);
    // only register if a cancelation frame exists
    cb.loc = hctx.hstack[top].loc1
    cb.loc.cancelers.push(cb);
    cb.cancel = f;
  }
  //return $std_core._unit_;
}

function _unregister_cancel( cb ) {
  if (typeof cb.cancel === "function") {
    cb.loc.cancelers = cb.loc.cancelers.filter( function(f) { return (f !== cb); } );
  }
  //return $std_core._unit_;
}



// --------------------------------------------------------
// Support for asynchronous actions with effect handlers
// --------------------------------------------------------

var _syncNesting = 0;
var _syncBlocked = null;

function $asyncStartSynchronous() {
  _syncNesting++;
  if (_syncBlocked===null) _syncBlocked = [];
}

function $asyncEndSynchronous() {
  if (_syncNesting===0) return;
  _syncNesting--; 
  if (_syncNesting===0) {
    // now invoke all queued callbacks; we schedule all through a timeout
    _syncBlocked.forEach( function(entry) {
      var cb = function() { return entry.fun.apply(entry.fun,entry.args); };
      setTimeout(cb,0);
    });
    _syncBlocked = null;
  }
}

// Create a callback wrapper around 'cb'. 
// If 'haserr' is true, the first parameter of the new call back is considered an error object
// and thrown automatically if not null. If it is null, it is not passed to the callback `cb`.
// If `nodup` is true, multiple call backs that are queued due to synchronized blocks are not duplicated
// and only the last call back is queued (for timers, mouse events etc).
function $asyncCallback( cb, haserr, nodup ) {
  // save the handler context
  var _hctx = $std_core._save_handler_context();
  // check for cancelation
  _check_canceled(_hctx);
  // remember if this callback itself was constructed inside a sync block
  var _in_synchronous = (_syncNesting > 0); 
  // and return a wrapped callback that restores the context
  var self = null;
  self = (function() {
    var args = Array.prototype.slice.call(arguments);
    // unregister cancelation
    _unregister_cancel(self);
    // check if we return while a 'synchronous' block was started
    if (!self.in_synchronous && _syncNesting > 0) { // if so, put ourselves on the queue
      if (nodup) {
        // update in-place possible previous occurrence of this event if 'nodup' was set (for timers or mouse events etc.)
        for (var i = 0; i < _syncBlocked.length; i++) {
          if (_syncBlocked[i].fun == self) {
            _syncBlocked[i].args = args;
            return; // updated in place, we are done.
          }
        }        
      }
      // and push ourselves on the queue to be called later
      _syncBlocked.push( { fun: self, args: args } );
      return;
    }
    else {
      // restore the handler context and invoke the callback
      $std_core._restore_handler_context( self.hctx, function(_) {
        if (haserr) {
          if (args[0] != null) throw args[0];
          args.shift();         
        }
        return cb.apply(cb,args);
      }, $std_core._unit_ );
    }
  });
  self.hctx = _hctx;
  self.in_synchronous = _in_synchronous;
  self.loc = null;
  self.cancel = null;
  self.canceled = false;
  return self;
}

function $asyncFork( action ) {
  setTimeout( _asyncCallback( function() {
    action($std_core.async_yield);
  }), 0 );  
}

function $asyncAll( actions, k ) {
  var n       = actions.length;
  var results = new Array(n);
  var exn;
  if (n<=0) return k(results);
  
  // all concurrent actions end with this continuation
  // only when all are finished, we resume the passed in continuation
  function allk(i,x,f) {
    results[i] = x;
    n--;
    if (n <= 0) {
      // if an exception was set, raise it now (or keep raising on multiple resume!)
      if (exn!==undefined) throw exn;
      // in case a branch exits more than once (due to amb for example)
      // note: this should never happen as multiple resumes can only happen after
      // at least one resumption but that cannot be past an asyncAll..
      for(i = 0; i < results.length; i++) {
        if (results[i] === undefined) return f(x); 
      }
      return k(results);
    }
    else {
      return f(x);
    }
  }

  for( var i = 0; i < n; i++) {
    (function(j){ 
      var cb = $asyncCallback( function(_) {
        return $std_core.$catch( function(k1) {
            return actions[j](k1);
          },
          function(_exn,_k2) {
            if (exn===undefined) exn = _exn;
            return allk(j, undefined, $std_core.async_yield);
          },
          function(x) { 
            return allk(j,x,$std_core.Result); 
          });
        }); 
      // by using setTimeout we can schedule all continuations for execution
      setTimeout(cb,0); 
    })(i);  // capture i by value
  }
  return $std_core.async_yield();
}
