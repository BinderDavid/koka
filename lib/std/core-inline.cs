/*---------------------------------------------------------------------------
  Copyright 2012 Microsoft Corporation.
 
  This is free software; you can redistribute it and/or modify it under the
  terms of the Apache License, Version 2.0. A copy of the License can be
  found in the file "license.txt" at the root of this distribution.
---------------------------------------------------------------------------*/
using System.Text;
using System.Collections.Generic;

public static class Primitive
{
  //---------------------------------------
  // Exceptions
  //---------------------------------------
  public static A Throw<A>( Exception exn ) {
    throw exn;
  }
  public static A Error<A>( string message ) {
    throw new ErrorException( message );
  }

  public static A PatternMatchError<A>( string location, string definition )
  {
    throw new PatternMatchException(location,definition);
  }

  public static A Unreachable<A>()
  {
    throw new ErrorException( "unreachable code reached");
  }

  public static A UnsupportedExternal<A>( string name ) {
    throw new ErrorException( "external '" + name + "' is not supported on this platform" );
  }

  public static A Catch<A>( Fun0<A> action, Fun1<Exception,A> handler )
  {
    try
    {
      return (A)action.Apply();
    }
    catch (Exception exn)
    {
      return (A)handler.Apply(exn);
    }
  }

  public static A Finally<A>( Fun0<A> action, Fun0<Unit> handler )
  {
    try
    {
      return (A)action.Apply();
    }
    finally
    {
      handler.Apply();
    }
  }

  //---------------------------------------
  // Run a stateful action safely
  //---------------------------------------
  public static A Run<A>( TypeFun1 action )
  {
    return (A)(((Fun0<A>)(action.TypeApply<Unit>())).Apply());
  }

  //---------------------------------------
  // Arrays
  //---------------------------------------
  public static A[] NewArray<A>( int len, A init )
  {
    A[] a = new A[len];
    for( int i = 0; i < len; i++) { a[i] = init; }
    return a;
  }

  public static A[] NewArray<A>( int len, Fun1<int,A> init )
  {
    A[] a = new A[len];
    for( int i = 0; i < len; i++) { 
      a[i] = (A)init.Apply(i); 
    }
    return a;
  }

  public static std_core._list<A> VList<A>( A[] v, std_core._list<A> tail ) {
    std_core._list<A> xs = tail;
    for(int i = v.Length-1; i >= 0; i--) {
      xs = new std_core._list<A>( v[i], xs );
    }
    return xs;
  }

  public static A[] UnVList<A>( std_core._list<A> xs ) {
    int len = 0;
    std_core._list<A> acc = xs;
    while(acc != std_core._list<A>.Nil_) { 
      len++;
      acc = acc.tail;
    }
    A[] v = new A[len];
    acc = xs;
    for(int i = 0; i < len; i++) { 
      v[i] = acc.head;
      acc = acc.tail;
    }
    return v;
  }


  //---------------------------------------
  // Dictionary
  //---------------------------------------
  public class Dict<T> : System.Collections.Generic.Dictionary<string,T> {
    public Dict() : base() {}
    public Dict( System.Collections.Generic.IDictionary<string,T> d ) : base(d) {}    
  }

  public static Dict<string> DictFromStringCollection( System.Collections.IDictionary d ) {
    Dict<string> dict = new Dict<string>();
    foreach( object key in d.Keys ) {
      if (key != null && key is string) {
        object val = d[key];
        if (val is string) {
          dict[(string)key] = (string)val;
        }
      }
    }
    return dict;
  }

  public class MDict<H,T> : System.Collections.Generic.Dictionary<string,T> {
    public MDict() : base() {}
    public MDict( System.Collections.Generic.IDictionary<string,T> d ) : base(d) {}
  }

  public static string[] DictKeys<A>( System.Collections.Generic.IDictionary<string,A> d ) {
    int i = 0;
    string[] result = new string[d.Keys.Count];
    foreach( string key in d.Keys) {
      result[i] = key;
      i++;
    }
    return result;
  }

  //---------------------------------------
  // Random
  //---------------------------------------
  private static Random random = new Random();
  public static double RandomDouble() 
  {
    return random.NextDouble();
  }

  public static int RandomInt()
  {
    return random.Next();
  }

  //---------------------------------------
  // Strings
  //---------------------------------------
  public static string Concat( string[] xs, string sep ) 
  {
    if (xs==null) return "";
    if (xs.Length==0) return "";
    StringBuilder sb = new StringBuilder(xs[0]);
    for (int i = 1; i < xs.Length; i++) {
      sb.Append(sep);
      sb.Append(xs[i]);
    }
    return sb.ToString();
  }

  public static int Count( string s, string pattern ) 
  {
    if (String.IsNullOrEmpty(pattern)) return 0;
    int count = 0;
    int i = 0;
    while( (i = s.IndexOf(pattern,i)) > 0 ) {
      count++;
    }
    return count;
  }

  public static string Repeat( string s, int n ) {
    if (n <= 0 || String.IsNullOrEmpty(s)) return "";
    StringBuilder sb = new StringBuilder("");
    for(int i = 0; i < n; i++) {
      sb.Append(s);
    }
    return sb.ToString();
  }

  public static int[] StringToChars( string s ) {
    List<int> v = new List<int>(s.Length);
    for(int i = 0; i < s.Length; i++) {
      v.Add( Char.ConvertToUtf32(s,i) );
      if (Char.IsHighSurrogate(s[i])) i += 1;
    }
    return v.ToArray();
  }

  public static int StringCount( string s ) {
    int n = 0;
    for(int i = 0; i < s.Length; i++) {
      n++;
      if (Char.IsHighSurrogate(s[i])) i+=1;
    }
    return n;
  }

  public static string CharToString( int c ) {
    return Char.ConvertFromUtf32(c);
  }

  public static string CharsToString( int[] v ) {
    StringBuilder sb = new StringBuilder();
    foreach( int c in v) {
      sb.Append( CharToString(c) );
    }
    return sb.ToString();
  }

  public static string Substr( string s, int start ) {
    return Substr(s,start,s.Length);
  }

  public static string Substr( string s, int start, int len ) {
    var idx = (start >= 0 ? start : s.Length + start);
    if (idx < 0) idx = 0;
    if (idx >= s.Length || len <= 0) return "";
    return (idx + len >= s.Length ? s.Substring(idx) : s.Substring(idx,len));
  }

  public static std_core._sslice SliceFirst( string s ) {
    if (String.IsNullOrEmpty(s)) 
      return new std_core._sslice("",0,0);
    else 
      return new std_core._sslice(s,0,Char.IsHighSurrogate(s[0]) ? 2 : 1);
  }

  public static std_core._sslice SliceLast( string s ) {
    if (String.IsNullOrEmpty(s)) 
      return new std_core._sslice("",0,0);
    else if (Char.IsLowSurrogate(s[s.Length-1])) 
      return new std_core._sslice(s,s.Length-2,2);
    else
      return new std_core._sslice(s,s.Length-1,1);
  }

  public static int SliceCount( std_core._sslice slice ) {
    int n = 0;
    for(int i = slice.start; i < slice.start + slice.len; i++) {
      n++;
      if (Char.IsHighSurrogate(slice.str[i])) i+=1;
    }
    return n;
  }

  public static std_core._sslice SliceExtend( std_core._sslice slice, int count ) {
    if (count==0) return slice;
    int i = slice.start + slice.len;
    if (count > 0) {
      while(i < slice.str.Length && count > 0) {
        count--;
        i += (Char.IsHighSurrogate(slice.str[i]) && i < slice.str.Length-1 ? 2 : 1);
      }
    }
    else {  
      while(i > slice.start && i > 0 && count < 0) {
        count++;
        i -= (Char.IsLowSurrogate(slice.str[i-1]) && i > slice.start+1 ? 2 : 1);
      }
    }
    return new std_core._sslice(slice.str, slice.start, (i > slice.start ? i - slice.start : 0));
  }

  public static std_core._sslice SliceAdvance( std_core._sslice slice, int count ) {
    if (count==0) return slice;
    int i   = slice.start;
    int end = slice.start + slice.len;
    int sliceCount = SliceCount(slice);
    int extra = 0;
    if (count > 0) {
      while(i < slice.str.Length && extra < count) {
        extra++;
        i += (Char.IsHighSurrogate(slice.str[i]) && i < slice.str.Length-1 ? 2 : 1);
      }
      if (end > i && sliceCount > extra) {
        return SliceExtend( new std_core._sslice(slice.str, i, end - i), extra );
      }
    }
    else {  
      while(i > 0 && extra < -count) {
        extra++;
        i -= (Char.IsLowSurrogate(slice.str[i-1]) && i > 1 ? 2 : 1);
      }
      if (sliceCount > extra) {
        return SliceExtend( new std_core._sslice(slice.str, i, slice.start-i), sliceCount - extra );
      }
    }
    return SliceExtend( new std_core._sslice(slice.str, i, 0), sliceCount );
  }


  public static string SliceToString( std_core._sslice slice ) {
    return slice.str.Substring(slice.start,slice.len);
  }

  //---------------------------------------
  // Trace
  //---------------------------------------
  public static void Trace( string msg ) 
  {
    System.Diagnostics.Debug.Print(msg);
    Console.Error.WriteLine(msg);
  }
  
  //---------------------------------------
  // ReadLine
  //---------------------------------------
  private static Async<string> onReadLine = null;

  public static Async<string> ReadLine()
  {
    if (onReadLine == null) onReadLine = new Async<string>();
    return onReadLine;
  }

  // For now, the MainConsole enters an event loop that handles
  // ReadLine from the Console. Later other asynchronous api's can
  // be added through the AsyncGlobal class, keeping the application
  // active as long as there are 'on' handlers installed.
  public static A MainConsole<A>( Fun0<A> f ) {
    A x = (A)f.Apply();
    while (!AsyncGlobal.AllDone()) {
      if (onReadLine != null) {
        string s = Console.In.ReadLine();
        Async<string> res = onReadLine;
        onReadLine = null;
        res.Supply(s);  // this may set onReadLine again
      }
      else {
        // bad
        Primitive.Trace("MainConsole: active async's but no readline");
        break;
      }
    }
    return x;
  }
};
  
//---------------------------------------
// Async
//---------------------------------------
public class AsyncGlobal 
{
  protected static int active = 0;

  public static bool AllDone() {
    return (active <= 0);
  }
}

public class Async<A> : AsyncGlobal
{
  Action<A> on = null;
  Action<Exception> onexn = null;
  bool done = false;
  Exception exn = null;
  A value;


  public bool IsDone {
    get { return done; }
  }

  public Async<B> On<B>( Fun1<A,B> fun ) {
    Async<B> result = new Async<B>();
    if (done) {
      if (exn == null) {
        result.Supply((B)fun.Apply(value));  
      }
    }
    else {
      if (on != null) {
        Action<A> prev = on;
        on = delegate(A x) { prev(x); result.Supply((B)fun.Apply(x)); };
      }
      else {
        active++;
        on = delegate(A x) { result.Supply((B)fun.Apply(x)); };
      }
    }
    return result;
  }

  public Async<B> OnExn<B>( Fun1<Exception,B> fun ) {
    Async<B> result = new Async<B>();
    if (done) {
      if (exn != null) {
         result.Supply((B)fun.Apply(exn));  
      }
    }
    if (onexn != null) {
      Action<Exception> prev = onexn;
      onexn = delegate(Exception x) { prev(x); result.Supply((B)fun.Apply(x)); };
    }
    else {
      active++;
      onexn = delegate(Exception x) { result.Supply((B)fun.Apply(x)); };
    }
    return result;
  }

  public void Supply( A x ) {
    if (done) return;
    done = true;
    value = x;
    if (on != null) {
      on(x);
      on = null; 
      active--;
    }
    if (onexn != null) {
      onexn = null;
      active--;
    }
  }

  public void SupplyExn( Exception x ) {
    if (done) return;
    done = true;
    exn = x;
    if (onexn != null) {
      onexn(exn);
      onexn = null; 
      active--;
    }
    if (on != null) {
      on = null;
      active--;
    }
  }

}


//---------------------------------------
// Exceptions classes
//---------------------------------------
public class ErrorException : Exception
{
  public ErrorException( string message ) : base(message) 
  {}
};

public class PatternMatchException : Exception
{
   public PatternMatchException( string location, string definition ) 
     : base( location + (String.IsNullOrEmpty(definition) ? "" : (": " + definition)) + ": pattern match failure" ) 
   { }
};

//---------------------------------------
// References
//---------------------------------------
public sealed class _Ref { }
public sealed class Ref<H,T> : TA<TA<_Ref,H>,T>
{
  public T Value;
  
  public Ref( T value ) {
    this.Value = value;
  }

  public void Set( T value ) {
    this.Value = value;
  }
}

//---------------------------------------
// Primitive types
//---------------------------------------

public enum Unit {
  unit
}

public interface TA<A,B>
{
}

public interface TypeFun1
{
  object TypeApply<A>();
}

public interface TypeFun2
{
  object TypeApply<A,B>();
}

public interface TypeFun3
{
  object TypeApply<A,B,C>();
}

public interface TypeFun4
{
  object TypeApply<A,B,C,D>();
}

public interface TypeFun5
{
  object TypeApply<A,B,C,D,E>();
}

public interface TypeFun6
{
  object TypeApply<A,B,C,D,E,F>();
}

public interface Fun0<in A> 
{
   object Apply();
}

public interface Fun1<in A, in B> 
{
  object Apply( A x );
}

public interface Fun2<in A1,in A2, in B> 
{
  object Apply( A1 x1, A2 x2 );
}

public interface Fun3<in A1, in A2, in A3, in B> 
{
  object Apply( A1 x1, A2 x2, A3 x3 );
}

public interface Fun4<in A1,in A2,in A3,in A4,in B> 
{
  object Apply( A1 x1, A2 x2, A3 x3, A4 x4 );
}

public interface Fun5<in A1,in A2,in A3,in A4,in A5, in B> 
{
  object Apply( A1 x1, A2 x2, A3 x3, A4 x4, A5 x5 );
}

public interface Fun6<in A1,in A2,in A3,in A4,in A5,in A6, in B> 
{
  object Apply( A1 x1, A2 x2, A3 x3, A4 x4, A5 x5, A6 x6 );
}
