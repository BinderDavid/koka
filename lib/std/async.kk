/*---------------------------------------------------------------------------
  Copyright 2012-2016 Microsoft Corporation.
 
  This is free software; you can redistribute it and/or modify it under the
  terms of the Apache License, Version 2.0. A copy of the License can be
  found in the file "license.txt" at the root of this distribution.
---------------------------------------------------------------------------*/

// Asynchronous primitives
module std/async

import std/num/int32
import std/time/duration

extern include {
  js file "async-inline.js"
}

type extend exception-info {
  Cancel
}

public fun cancelable( action : (cancel:() -> exn ()) -> <async,exn|e> a ) : <async,exn|e> maybe<a> { 
  val caction = create-cancelable-prim(action)
  catch { Just(caction()) } fun(exn) {
    match(exn.info) {
      Cancel -> {
        check-canceled() // will re-throw if action is not the one that was canceled
        Nothing
      }
      _ -> throw(exn)
    }
  }
}

extern create-cancelable-prim( action : (cancel:() -> exn ()) -> <async,exn|e> a ) : (() -> <async,exn|e> a) {
  js "_create_cancelable"
}

extern check-canceled() : exn () {
  js "_check_canceled"
}

public extern unsafe-register-cancel( cb : a -> <async|e> (), cancel : () -> exn () ) : exn () {
  js "_register_cancel"
}

public fun timeout( secs : duration, action : () -> <async,exn|e> a ) : <async,exn|e> maybe<a> {
  cancelable fun(cancel) {
    val h = set-timeout( fun(_){ cancel(); }, secs.milli-seconds.int32 )
    on-exit({ cancel-timeout(h) }, action)
  }
}

// ----------------------------------------------------------------------------
// Asynchronous operations are supported through the `:async` effect.
// ----------------------------------------------------------------------------

public fun async-fork( action1 : () -> <async|e> a, action2 : () -> <async|e> ()) : <async|e> a {
  val xs = async-all( [{ Just(action1()) }, { action2(); Nothing }] )
  unsafe-noexn{ xs.head.unjust.unjust }
}

// Execute `actions` asynchronously, i.e. interleaved around asynchronous operations.
// When all operations have terminated (or raised an exception) return the list of
// results (or raise the first raised exception).
public fun async-all( actions : list<() -> <async|e> a> ) : <async|e> list<a> {
  async-all-prim( actions.vector ).list;
}


extern async-all-prim( actions : vector<() -> <async|e> a> ) : <async|e> vector<a>  {
  js inline "$asyncAll(actions,_k)"
}

extern async-callback0( action: () -> e (), haserr : bool, nodup : bool ) : (() -> e ()) {
  js "$asyncCallback"
}

public fun unsafe-async-yield-err( action : (cb : a -> <async|e> ()) -> <async|e> () ) : <async|e> a {
  unsafe-async-yield( action, True ) 
}

public extern unsafe-async-yield( action : (cb : a -> <async|e> ()) -> <async|e> (), haserr : bool = False ) : <async|e> a {
  js inline "(#1)( $asyncCallback(_k,#2), $std_core.async_yield)"
}

// Wait (asynchronously) for `secs` seconds. 
// Use `wait(0)` to yield to other asynchronous operations.
public fun wait( secs : double ) : <async,exn> () {
  wait(secs.duration)
}

// Wait (asynchronously) for `secs` seconds duration. 
// Use `wait(0)` to yield to other asynchronous operations.
public fun wait( secs : duration ) : <async,exn> () {
  unsafe-async-yield( fun(cb) { 
    val msecs = secs.milli-seconds.int32
    val h = set-timeout( cb, msecs)
    if (msecs > 100.int32) { 
      unsafe-register-cancel( cb, fun(){ cancel-timeout(h) } )
    }
  })
}

type timeout-handle

extern set-timeout( action : (()) -> e (), ms : int32 ) : exn timeout-handle {
  js "setTimeout"
}

extern cancel-timeout( h : timeout-handle ) : exn () {
  js "clearTimeout"
}


// Execute any asynchronous actions inside `action` synchronously, ie. 
// no other actions are interleaved.
public fun synchronous( action : () -> <async|e> a ) : e a {
  start-synchronous()
  on-exit(end-synchronous){ unsafe-total(action) }
}

extern start-synchronous() : () {
  js "$asyncStartSynchronous"
}

extern end-synchronous() : () {
  js "$asyncEndSynchronous"
}

