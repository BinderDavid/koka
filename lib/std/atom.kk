/*---------------------------------------------------------------------------
  Copyright 2012 Microsoft Corporation.
 
  This is free software; you can redistribute it and/or modify it under the
  terms of the Apache License, Version 2.0. A copy of the License can be
  found in the file "license.txt" at the root of this distribution.
---------------------------------------------------------------------------*/

/* This module provides a perfect hash for strings.

   These `:atom`s can be compared very efficiently. If you compare only
   seldomly, a straight string compare will probably be faster since 
   creating an atom takes some initial work.
*/
module std/atom

import std/vector

external include {
  cs file "atom-inline.cs"
  js file "atom-inline.js"
}

// An `atom` corresponding to the empty string
public val atom-null = atom("")

// An atom is an abstract representation of a string that can be compared quickly
abstract struct atom( value: int )  

// Create an `:atom`
extern atom-create : ( s : string ) -> int {
  cs "Atom.Create"
  js "$atomCreate"
}

// Convert an `:atom` back to a `:string`
extern atom-to-string : ( x : int ) -> string {
  cs "Atom.ToString"
  js "$atomToString"
}

// Create an `:atom`
public fun atom( s : string ) : atom {
  atom-create(s).Atom
}

// Convert an `:atom` back to a `:string`
public fun string( a : atom ) : string {
  atom-to-string(a.value)
}

// Compare two atoms for equality
public fun (==)(x : atom, y : atom) : bool  -> (x.value == y.value)

// Compare if two atoms are not equal.
public fun (!=)(x : atom, y : atom) : bool  -> (x.value != y.value)


extern atom-matcher : ( xs : vector<string>, ignoreCase: int ) -> any
{
  cs "Atom.CreateMatcher"
  js "$atomCreateMatcher"
}

extern atom-match : ( m : any, s : string, ignoreCase: int ) -> int
{
  cs "Atom.Match"
  js "$atomMatch"
}

// Create an efficient string matcher which given a list of strings returns a matching function.
public fun matcher( xs : list<string>, ignoreCase: bool = False ) : (( s : string ) -> bool)
{
  matcher(vector(xs), ignoreCase)
}

// Create an efficient string matcher which given a vector of strings returns a matching function.
public fun matcher( xs : vector<string>, ignoreCase: bool = False ) : (( s : string ) -> bool)
{
  val m = atom-matcher( xs, ignoreCase.int );
  return fun(s) { return bool(atom-match(m,s,int(ignoreCase))); }
}
