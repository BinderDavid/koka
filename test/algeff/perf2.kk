import std/date

function timer(f) {
  val t0 = now().time
  val adjust = now().time - t0
  println("start")
  val start = now().time
  val x = f()
  val end = now().time
  println("ended in " + (end - start).show + "ms")
  x
}

alias solution = list<int>
alias solutions = list<list<int>>

function safe( queen : int, diag : int, xs : solution ) : bool
{
  match(xs) {
    Cons(q,qs) -> (queen != q && queen != (q+diag) && queen != (q - diag) && safe(queen,diag+1,qs))
    _          -> True
  }
}

function appendSafe( queen : int, xs : solution, xss : solutions ) : div solutions
{
  if (queen <= 0) xss
  else {
   if (safe(queen,1,xs))
    then appendSafe( queen-1, xs, Cons(Cons(queen,xs),xss) )
    else appendSafe( queen-1, xs, xss )
  }
}

function extend(queen : int, acc : solutions, xss : solutions ) : div solutions
{
  match(xss) {
    Cons(xs,rest) -> extend(queen, appendSafe(queen,xs,acc), rest)
    Nil           -> acc
  }
}

function findSolutions( n : int, queen : int ) : div solutions
{
  if (queen == 0) 
   then [[]]
   else extend(n,[], findSolutions(n,queen-1))  
}

public function queens( n : int ) : div int
{
  length(findSolutions(n,n))
}


effect choose {
  pick(n : int) : int;
  fail<b>()     : b
}

function listmap(n : int, f : int -> e a) : e list<a> {
  if (n < 1) then [] else Cons(f(n),listmap(unsafeDecreasing(n-1),f))
}

function chooseAll( action ) {
  val h = handler{
    return x  -> [x]
    pick(n)   -> {
      list(1,n).map(resume).concat
    }
    fail      -> []
  }
  h(action)
}

function extend1( n : int, xs : solution ) : choose solution {
  val i = pick(n)
  if (safe(i,1,xs)) then Cons(i,xs) else fail()
}

function findSolution( n : int, queen : int ) : choose solution {
  if (queen==0) then []
   else extend1(n, findSolution(n, unsafeDecreasing(queen-1)))
}

function showSolutions( xss : solutions ) : io () {
  xss.foreach fun(xs) {
    println(xs.show)
  }
}

function queensChoose(n : int ) : io () {
  val xss = chooseAll({findSolution(n,n)} )
  xss.showSolutions
}

function test(n) {
  println( queens(n).show )
  queensChoose(n) 
}

/*
safeAddition :: [Int] -> Int -> Int -> Bool
safeAddition [] _ _ = True
safeAddition (r:rows) row i =
   row /= r &&
   abs (row - r) /= i &&
   safeAddition rows row (i + 1)

-- hand-coded solution to the n-queens problem
queensPure :: Int -> [[Int]]
queensPure n = foldM f [] [1..n] where
    f rows _ = [row : rows |
                row <- [1..n],
                safeAddition rows row 1]

-- n queens using an abstract Choose operator
queensComp :: [handles|h {Choose}|] => Int -> Comp h [Int]
queensComp n = foldM f [] [1..n] where
    f rows _ = do row <- choose [1..n]
                  if (safeAddition rows row 1)
                    then return (row : rows)
                    else failed
*/