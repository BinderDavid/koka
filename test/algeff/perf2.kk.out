.Eff-choose : forall<a> (.ops-choose<a>) -> operation<a>
.Op-fail : forall<a,b> .ops-choose<a>
.Op-pick : forall<a> (int) -> .ops-choose<a>
appendSafe : (queen : int, xs : solution, xss : solutions) -> div solutions
chooseAll : forall<a,e> (action : () -> <choose|e> a) -> e list<a>
extend : (queen : int, acc : solutions, xss : solutions) -> div solutions
fail : forall<a> () -> choose a
findSolution: (n : int, queen : int) -> <choose,div> solution
findSolutions: (n : int, queen : int) -> div solutions
main : () -> <div,console,ndet> ()
pick : (n : int) -> choose int
queens : (n : int) -> div solutions
queensChoose: (n : int) -> div solutions
safe : (queen : int, diag : int, xs : solution) -> bool
showSolutions: (xss : solutions) -> io ()
test : (n : int) -> <div,console,ndet> ()
timer : forall<a,e> (f : () -> <console,ndet|e> a) -> <console,ndet|e> a